File: spec/filetypes_spec.rb
# frozen_string_literal: true

require 'spec_helper'

RSpec.describe 'filetypes' do
  it 'creates mapping for bullets on empty buffer if configured' do
    vim.command 'new'
    vim.insert '# Hello there'
    vim.feedkeys '\<cr>'
    vim.type '- this is the first bullet'
    vim.feedkeys '\<cr>'
    vim.type 'this is the second bullet'

    buffer_content = vim.echo("join(getbufline(bufname(''), 1, '$'), '\n')")

    expect(buffer_content).to eq normalize_string_indent(<<-TEXT)
      # Hello there
      - this is the first bullet
      - this is the second bullet
    TEXT
  end

  it 'should have text filetype for .txt' do
    # bullets.vim is triggered by particular filetypes;
    # if somehow your vim is recognising .txt and setting
    # filetype to something that isn't text or markdown,
    # the rest of the tests are gonna fail.
    filename = "#{SecureRandom.hex(6)}.txt"
    write_file(filename, '')

    vim.edit filename
    vim.type 'i'
    vim.feedkeys '\\<c-r>'
    vim.type '=&filetype'
    vim.feedkeys '\\<cr>'
    vim.write

    file_contents = normalize_string_indent(IO.read(filename)).strip

    expect(%w[markdown text]).to include(file_contents)
  end
end

File: test/filetypes_spec.lua
local helpers = require("test.helpers")

describe("filetypes", function()
  it("creates mapping for bullets on empty buffer if configured", function()
    -- g:bullets_enable_in_empty_buffers defaults to 0, so a new buffer with
    -- no filetype does NOT get the bullet <CR> mapping.
    vim.cmd("enew")
    vim.cmd("setlocal formatoptions= comments=") -- prevent '#' comment-continuation
    helpers.feedkeys("i# Hello there<CR>- this is the first bullet<CR>this is the second bullet<Esc>")
    assert.are.same({ "# Hello there", "- this is the first bullet", "this is the second bullet" }, helpers.get_lines())
  end)

  it("should have text filetype for .txt", function()
    local tmpfile = vim.fn.tempname() .. ".txt"
    vim.cmd("edit " .. tmpfile)
    local ft = vim.bo.filetype
    assert.is_true(ft == "text" or ft == "markdown", "Expected 'text' or 'markdown' filetype, got: " .. tostring(ft))
    vim.cmd("bdelete!")
  end)
end)
